docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,116 @@
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSHealthCheck
+metadata:
+  name: example-dnshealthcheck
+  ownerReference:
+    type: DNSPolicy
+    name: example-policy
+spec:  
+  config:
+    port: "..."
+    path: "..."
+    protocol: "..."
+    host: "..."
+    ipAddress: "..."
+  provider: "aws | agent | ..."
+  providerSpecific:
+    "agent/managed-cluster": "..."
+status:
+  providerSpecific:
+    "agent/health-status": "Healthy"
+    "agent/last-check": "..."
+    "aws/health-check-id": "..."
+```
+
+Relevant fields
+
+| Field name | Description |
+| ---------- | ----------- |
+| `metadata.ownerReference` | Reference to the owner of the health check. Keeps track in order to propagate deletion and reconciliation. An update to the DNSHealthCheck should trigger the reconciliation of the owner resource |
+| `spec.config` | Provider agnostic configuration section of the health check |
+| `spec.provider` | Name of the health check provider to use |
+| `spec.providerSpecific` | Provider specific configuration |
+| `status.providerSpecific` | Provider specific status |
+
+The reconciliation logic of this resource broadly follows:
+
+1. Select the health check reconciler implementation based on the value of the provider field
+1. Delegate the reconciliation on the health check reconciler, passing the configuration and the provider specific values
+1. Update the status and the providerSpecific status values from the reconcile result
+1. The provider specific status values are propagated to the related DNSRecord
+
+##### Providers
+
+The `.spec.provider` field allows the selection of the provider that reconciles the
+health checks. The following providers are proposed:
+
+| Name | Description |
+| ---- | ----------- |
+| `aws` | Supports the current health check implementation. It creates or updates a health check in Route53 configured with the values of `.spec.config`. Sets the `aws/health-check-id` provider specific field |
+| `agent` | Places a `DNSHealthCheckProbe` resource in the managed cluster that serves the endpoint. The status of the `DNSHealthCheckProbe` is propagated into the `agent/<status field>` provider specific status. |
+
+### Data plane
+
+#### `DNSHealthCheckProbe` resource
+
+The DNSHealthCheckProbe resource configures the agent in the data plane cluster to perform the health checks against a local endpoint.
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSHealthCheckProbe
+metadata:
+  name: example-probe
+spec:
+  port: "..."
+  host: “...”
+  ipAddress: "..."
+  path: "..."
+  protocol: "..."
+  interval: "..."
+status:
+  status: Healthy
+  lastCheck: "..."
+```
+
+The reconciliation of this resource results in the deployment of the health check agent, which performs health checks following the configuration in the spec, and updates the status. The status is propagated upstream to the providerSpecific status of the equivalent DNSHealthCheck
```

Comment:
Would the agent not already have been placed on the workload cluster prior to this? Do you mean that the control plane will ensure it's deployed on cluster when we first know we want to add an agent health probe?